### PR TITLE
Fix -profile-ltac flag

### PR DESCRIPTION
**Pull request by JasonGross**
This is probably not the Right way to fix it, but I don't know how to tell Coq that it shouldn't ignore side effects of tactics on the document state when in profiling mode.  (@gares? @ppedrot?)

Even if the fix for the flag itself doesn't get merged, the comment-only changes in the debug-printing should probably get merged, and the testing file should also probably be merged.  (This is on top of #217 because I need the support I added there to testing infrastructure.)

This probably doesn't need any kind of copyright change, since I'm doing this on my own time on my own machine.

### CODE REVIEW

**Comment by JasonGross** [comment]
Bump.  Any update on this?

**Comment by JasonGross** [comment]
Bump.  This should probably get the 8.6 tag, since -profile-ltac is useless without this patch.

**Comment by ppedrot** [comment]
I don't really understand what is going on here, as I'm not a user of the profiler (and the initial code is ugly anyway). Can we trust you on this one?

**Comment by JasonGross** [comment]
The code works, and it's better than nothing, given that the current state of things is "`-profile-ltac` claims that no tactics were run and that the total time spent in tactics is 0".   Is that enough to get it merged?  (Once this is merged, I'll probably be making use of `-profile-ltac` on every build, so hopefully I'll catch any remaining wrinkles in the results.)



The issue is that `Defined` and `Qed` reset the document state to the way it was before the proof started, add the proof to the global env, and then replay any vernaculars declared as side-effectful (or something similar; I'm sure @gares has a better way of describing this).  Since tactics are not side-effectful, this means that every `Defined` and `Qed` wipe the profiling data that was gathered during execution of the tactic script.

To get around this, I make a copy of the profiling data stack which is not synchronized with the document, and update this one in parallel with the synchronized one.  (The unsynchronized one is used only for the `-profile-ltac` flag, currently, while the synchronized one is used to display things during document execution.  This is not great, but it's better than what's there now, and I'm waiting on feedback from @gares (or anyone else), on how to synchronize the profiling data with backtracking, but not wipe it when proofs are closed.)

Since @tebbi makes use of imperative update of structures with multiple pointers to them, I took some care to make sure that the root node of the profiling data is shared between the synchronized and unsynchronized versions in `reset_profile` (but perhaps this is not necessary, since the code works, and I think sharing is lost when the state is restored), and I also take some care to guard against double-counting when references are shared, by checks of physical equality.

The rest of the changes at the bottom are just factoring the printing methods over which profile data to display.

**Comment by ejgallego** [comment]
Hey @jason, would it help for bactracking if you associated the profiling data to the current `StateId.t` ?

**Comment by JasonGross** [comment]
@ejgallego Are there appropriate backtracking hooks?  Can you sketch out your idea a bit more?  (The part that I currently have is a type for "profiling data so far" and a box that gets invoked on tactic execution and knows how to (imperatively) update something of type "profiling data so far" with the profiling data from the current tactic.  The way this is currently handled is that there's a `Summary.ref` cell holding the "profiling data so far".  This works well with backtracking, but, unfortunately, when the first tactic is invoked after a `Defined`/`Qed`, the data in the `Summary.ref` cell is that associated with the state before the previous proof was opened, rather than the state right before the previous proof was closed.)

**Comment by ejgallego** [comment]
Hi Jason, what I was proposing was something more in the line of not using `Summary.ref` to store the data but have a profile map indexed by `Stateid.t`. Then, you could reconstruct any profile info by following the DAG.

Unfortunately I am barely familiar with the Stm so I'd wait for feedback from more knowledgeable people before implementing anything.

**Comment by gares** [comment]
The reason why get this seemingly odd behavior is that Proof.Qed blocks are modeled having the worst case scenario in mind.  Imagine the block is run by a worker: you can't have at line 9 such profiling data, since it is still being produced.

I think the cleanest solution is to stream profiling data via feedback and aggregate it on the UI side.
Alternatively, don't use Summary.ref but a simple ref, and as Emilio proposes store each piece of data together with its state id.  Then, when the user asks for a printer report, filter the data keeping only state_ids that are still valid.

**Comment by gares** [comment]
Note that "UI side" could be coqtop/coqc as well, by setting an appropriate feedback listener.  This is the only was to aggregate reliably data when things happen in parallel.

**Comment by JasonGross** [comment]
@gares I'm not sure what the aggregation on the coqtop UI side would look like.  I guess the code would mostly be the same, but I'd send a feedback message in a hook registered with (is it an invariant that for any tactic sentence included in the document, `Stm.get_current_state` returns the same state from the time `Stm.tactic_being_run_hook` is invoked with `true` to the time it's invoked with `false`?)  I think I'd rather leave the streaming to someone more familiar with the IDEs / messages, and implement @ejgallego's suggestion.  Would it go something like this?

- `Map.Make(Stateid.Self)` to map `Stateid.t` to profiling data
- Each time we save profiling data, associate it to `Stm.get_current_state ()` in the map
- Each time we get a "reset", note that the `Stm.get_current_state ()` has a reset associated with it
- On display, filter the "resets" for ones for which `Stm.get_current_state ()` is `Stateid.newer_than` them and also for which `Stm.state_of_id` says `Valid`, possibly minimize the set of resets by removing any `Stateid.t` `a` for which we have a `b` with `Stateid.newer_than b a`, and finally aggregate the profiling data from all states that both `Stm.state_of_id` says are `Valid` and which are `Stateid.newer_than` all resets that we care about.

(I'll try to implement this in the next two or three weeks, when I get some time to myself (since I don't want to do it on Google's machines, since I'm still waiting to hear back from @maximedenes about copyright / CLIs for things done on Google's time / Google's hardware).  Alternatively, if @ppedrot or @psteckler or @tebbi (or anyone else) wants to do this, that'd be great.)

**Comment by ejgallego** [comment]
Hey @JasonGross , that looks very interesting and I'd be interested to help or at least to figure out how the Stm should interact with your use case, however I'll be on holidays for the next few weeks,

I'd start indeed trying to save the profile data for each state i.e. node of the document. Then provide an ML API to get such information. This is not perfect because states may become invalid, but could be a first start.

Regarding the user-side aggregation, for instance in SerAPI you would do something like
```
(Query ((IdRange ($id1 id2)) ProfileData)
```
(Note that `(Query () ProfileData)` works today)

This would give you the corresponding profile data for a particular sentence, for instance to show on hover; of course you could also aggregate.

Cheers,
E.

**Comment by psteckler** [comment]
If that's supported on the ML side, of course, PG could use the XML
protocol to obtain that information.

-- Paul

**Comment by gares** [comment]
I mean something quite different with aggregation on the UI side.  Imagine you enable profiling on 3 sentences.  Coq is simply programmed to send a feedback message for each sentence with the data relative to the sentence.  The UI receives such data and prints it.  It could sum up the data of the 3 sentences as you do today, or give a more fine grained presentation on each sentence, or draw a flame graph... a basic UI like coqc could simply intercept all profiling feedback and print it at exit.  Coq (the non UI part) could avoid storing the data and just send out feedback messages is ltac profiling is enabled.

What Emilio proposes for serapi requires more work on the Coq side (storing and filtering data) but is doable.  Let me try to clarify the 2 options we have (in general, not only in this case).

I thing some class of meta data should be sent out by Coq (not stored, nor processed) and processed in the frontend (coqc, jscoq, CoqIDE...).   In particular for debugging features with a natural scope delimiter in the proof text.  Another example is hyperlinks, one does not want the UI to query for them, since they are almost always useful and Coq has no use for them, so such data is not even stored (it only exists transiently).
For other kind of metadata, an explicit query, with no change to the text buffer, I prefer the explicit query Emilio proposes.  For example one could reply to such a query by re-running some sentences with more flags turned on.  Imagine a "show me the intermediate goals of a multi rule rewrite tactic".  One would never want to alter the text to access such piece of info, and systematically generating and storing it is too heavy. Some other metadata is already stored, or easily re-generated, and rarely useful.  Again an explicit query fits the bill.

To sum up: some meta-data is naturally stored in Coq, some is not or is impossible to produce or store efficiently, some is always useful some other is rarely needed.  Depending on the specific case
an explicit query or an automatic production of feedback may be best.

At the light of such distinction, for ltac profiling I imagine one can live with writing StartLtacProf...StopLtacProf, store no data on the Coq side and let the UI to most of the work.
Of course the main motivation of the subtle distinction, any my preference, is to leave as much code as possible out of Coq... that you may disagree with ;-)

**Comment by ejgallego** [comment]
Hi Enrico, well if I understand the code correctly for this case both approaches (sending vs querying) make little difference on the Coq side, if you want to support the vernacular for the toplevel.

I usually prefer the `Query` approach as it seems a bit heavy to me to send everything to the ide, profile data could become very large, programs too. Aggregation of course should be handled in SerAPI not in Coq.

It also forces it to handle messages it may not be prepared for, making backwards compatibility harder, compared to coq just not sending anything if not requested.

The point you make about the hyperlinks (I assume you mean Glob info) is an interesting one, are they always generated at parsing time?

**Comment by Ptival** [comment]
Hmm sorry, I'm not quite sure I understand fully the situation but if does seem like we'd rather want to query for profiling data than get it by default.

But that's only my two cents having skimmed the entire conversation. It would help to have an idea of the format and possible size of such data. :-)

**Comment by gares** [comment]
@ejgallego glob info is not stored in therms (constr).  In particular terms lose all links with their original position in the text, only the AST has such loc.  Glob info (hyperlink) needs the loc.  So such piece of info is only transiently available, when you "internalize" (resolve names) the AST (expr_AST --intern-->glob_AST --interp--> constr).

Wrt this specific "query", given that you delimit the text to be profiled explicitly (sneaking a command before and one after), requiring an extra query to access such data is a bit weird to me.  E.g. if you query outside the range?  Anyway, I'm fine with any solution.

Another thing, ignoring unrecognized feedback is way to be forward compatible.  So it must be supported by UIs.  Also, a Coq plugin could emit his own feedback and provide extra functionality to IDE1, so IDE2 must be resilient to unknown feedbacks.

**Comment by cpitclaudel** [comment]
I don't have a strong opinion either way, but we should be mindful of excessive message-passing. Profiles might be large, and sending them over will take time.

**Comment by ppedrot** [comment]
Isn't this PR deprecated since commit ebb1048b?

**Comment by JasonGross** [comment]
Indeed, this PR is deprecated, but can we get 20a474f and 03adb72 into the v8.6 branch, so we can test profiling output format?

**Comment by JasonGross** [comment]
I could make a new PR with those, if that's preferred.
